library/std/src/panicking.rs

```
@@ -336,20 +336,26 @@ pub mod panic_count {
     pub fn increase() -> (bool, usize) {
         let global_count = GLOBAL_PANIC_COUNT.fetch_add(1, Ordering::Relaxed);
         let must_abort = global_count & ALWAYS_ABORT_FLAG != 0;
-        let panics = if must_abort {
-            global_count & !ALWAYS_ABORT_FLAG
-        } else {
-            LOCAL_PANIC_COUNT.with(|c| {
+        let mut panics = global_count & !ALWAYS_ABORT_FLAG;
+
+        // In a restricted_std environment, we likely don't have thread_local.
+        // If we attempt to use LOCAL_PANIC_COUNT here, it is probable that another panic will
+        // occur, sending us into an infinite panic loop that never calls the panic handler.
+        #[cfg(not(feature = "restricted-std"))]
```

Rather than ignoring recursive panics, maybe set `must_abort` to true?